apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PriorityList.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// A simple sorted list that allows items to return equality for IComparable.CompareTo
+    /// </summary>
+    internal class PriorityList<T> where T : IPriorityListItem<T>
+    {
+        private readonly List<T> _innerList = new List<T>();
+
+        public int Count
+        {
+            get
+            {
+                lock (_innerList)
+                {
+                    return _innerList.Count;
+                }
+            }
+        }
+
+        public IEnumerable<T> GetSnapshot()
+        {
+            lock (_innerList)
+            {
+                return _innerList.ToArray();
+            }
+        }
+
+        public void Add(T item)
+        {
+            lock (_innerList)
+            {
+                var index = _innerList.BinarySearch(item);
+                if (index < 0)
+                {
+                    //if priority changes during this call, it will be blocked from
+                    //processing until after Add has completed.
+                    item.PriorityChanged += Item_PriorityChanged;
+
+                    _innerList.Insert(~index, item);
+                }
+            }
+        }
+
+        public void Remove(T item)
```

Comment:
Why can't this use `BinarySearch` like adding can? Is it because at the time the `PriorityChanged` event is handled, the item already has the new priority, so it can't be found here? One way to solve that might be to also have a `PriorityChanging` event and you remove on changing and re-add on changed, but I'm not sure how that might affect the locking logic. You could also have the `PriorityChanged` event include the old priority in the event args, but then binary search is a bit messier since `BinarySearch` needs an item type. You'd need to create a dummy item to hold the old priority or re-implement binary search manually (since you had that logic already, perhaps it isn't a bad option).